docs/source/etcd-architecture.rst

```diff
@@ -0,0 +1,238 @@
+~~~~~~~~~~~~~~~
+
+The orchestrator will inevitably have its own set of APIs for managing
+networks. The orchestrator plugin's primary job is to translate those APIs into
+the Calico etcd data model (see :doc:`etcd-data-model`) to allow Calico to
+perform the appropriate network programming.
+
+Some of this translation will be very simple, other bits may be more complex in
+order to render a single complex operation (e.g. live migration) into the
+series of simpler operations the rest of the Calico network expects.
+
+Feedback
+~~~~~~~~
+
+If necessary, the orchestrator plugin will provide feedback from the Calico
+network into the orchestrator. Examples include: providing information about
+Felix liveness; marking certain endpoints as failed if network setup failed.
+
+
+.. _calico-etcd-component:
+
+etcd
+----
+
+Calico uses etcd as its backing data store. etcd is a distributed key-value
+store that has a focus on consistency, ensuring that Calico can always build an
+accurate network.
+
+In addition to its role as Calico's primary data store, etcd also acts as a
+communication mechanism between the various components. We do this by having
+the non-etcd components watch certain points in the keyspace to ensure that
+they see any changes that have been made, allowing them to respond to those
+changes in a timely manner.
+
+The etcd component is distributed across the entire deployment. It is divided
+into two groups of machines: the core cluster, and the proxies.
+
+In Calico, we deploy an etcd cluster. In small deployments this can be an etcd
+cluster of one, which provides no redundancy but is simple and low-overhead.
+In larger deployments we scale this up, as per the `etcd admin guide`_.
+
+Additionally, on each machine that hosts either a
+:ref:`calico-felix-component` or a :ref:`calico-orchestrator-plugin`, we run
+an etcd proxy. This is an attempt to reduce load on the core cluster and to
+ensure that nodes are shielded from the specifics of the etcd cluster.
+
+etcd is responsible for performing all of the following tasks:
+
+.. _etcd admin guide: https://github.com/coreos/etcd/blob/master/Documentation/admin_guide.md#optimal-cluster-size
+
+Data Storage
+~~~~~~~~~~~~
+
+etcd stores the data for the Calico network in a distributed, consistent,
+fault-tolerant manner. This set of properties ensures that the Calico network
+is always in a known-good state, while allowing for some number of the machines
+hosting etcd to fail or become unreachable.
+
+This distributed storage of Calico data also improves the ability of the Calico
+components to read from the database (which is their most common operation), as
+they can distribute their reads around the cluster.
+
+Communication
+~~~~~~~~~~~~~
+
+etcd is also used as a communication bus between components through the
+mechanism of having various components watch keys for changes. This allows the
+act of committing state to the database to cause that state to be programmed
+into the network.
```

Comment:
Wait, where did I say it? Because it'd be better said here in a section that is easy to find, than said somewhere else scattered in the prose.